external/fv3fit/fv3fit/reservoir/domain.py

```diff
@@ -8,3 +14,147 @@ def slice_along_axis(arr: np.ndarray, inds: slice, axis: int = 0):
     sl = [slice(None)] * arr.ndim
     sl[axis] = inds
     return arr[tuple(sl)]
+
+
+class RankDivider:
+    def __init__(
+        self,
+        subdomain_layout: Layout,
+        rank_dims: Sequence[str],
+        rank_extent: Sequence[int],  # shape of full data, including overlap
+        overlap: int,
+    ):
+        """ Divides a rank of data into subdomains for use in training.
+        Args:
+            subdomain_layout: tuple describing subdomain grid within the rank
+            rank_dims: order of dimensions in data. If using time series data, 'time'
+                must be the first dimension.
+            rank_extent: Shape of full data. This includes any halo cells from
+                overlap into neighboring ranks.
+            overlap: number of cells surrounding each subdomain to include when
+                taking subdomain data.
+
+        Ex. I want to train reservoirs on 4x4 subdomains with 4 cells of overlap
+        across subdomains. The data is preprocessed and saved as 1 C48 tile per rank,
+        with n_halo=4. I would initialize the RankDivider as
+            RankDivider(
+                subdomain_layout=(12, 12),
+                rank_dims=["time", "x", "y", "z"],
+                rank_extent=[n_timesteps, 56, 56, 79],
+                overlap=4,
+            )
+        """
+        self.subdomain_layout = subdomain_layout
+        if "time" in rank_dims:
+            if rank_dims[0] != "time":
+                raise ValueError("'time' must be first dimension.")
+        if not {"x", "y"}.issubset(rank_dims):
+            raise ValueError(
+                "'x' and 'y' dims must be in the rank_dims of the RankDivider"
+            )
+        self.rank_dims = rank_dims
+        self.overlap = overlap
+        self.rank_extent = rank_extent
+
+        self.x_ind = rank_dims.index("x")
+        self.y_ind = rank_dims.index("y")
+
+        self._partitioner = pace.util.TilePartitioner(subdomain_layout)
+        self.n_subdomains = subdomain_layout[0] * subdomain_layout[1]
```

Comment:
nit (non-blocking): It seems a little random here which variables are private and public. It may make things clearer to make all private variables private.